src/datastore.rs

```diff
@@ -0,0 +1,561 @@
+/// If initialized from disk, data can be loaded eagerly or lazily, as in, on access.
+/// It will remember the root data directory for this purpose. This complicates the
+/// accessor methods somewhat, because 1. access can fail with an IO error and 2.
+/// insertion can fail. Data is wrapped in a [`std::sync::Arc`] to help on-demand loading.
+///
+/// Note that it tracks files, not directories. Data paths you insert must not have
+/// any existing path in the store as an ancestor, or you would nest a file under a
+/// file. Images must always be in a flat directory. The paths are always relative to
+/// a UFO's data directory.
+///
+/// # Example
+///
+/// Consider a UFO on disk with the following structure:
+///
+/// ```text
+/// Test.ufo/
+/// ├── data/
+/// │   ├── a.txt
+/// │   ├── b.bin
+/// │   └── com.testing.random/
+/// │       ├── c.txt
+/// │       └── zzz/
+/// │           └── z.txt
+/// ├── images/
+/// │   ├── image1.png
+/// │   ├── image2.png
+/// │   └── image3.png
+/// ├── glyphs/
+/// │   ├── a.glif
+/// │   └── contents.plist
+/// ├── layercontents.plist
+/// └── metainfo.plist
+/// ```
+///
+/// The `data` subfolder will be represented in a [`DataStore`] like so:
+///
+/// * `PathBuf::from("a.txt")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("b.bin")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("com.testing.random/c.txt")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("com.testing.random/zzz/z.txt")` → `b"<content>".to_vec()`
+///
+/// The `images` subfolder will be represented in an [`ImageStore`] like so:
+///
+/// * `PathBuf::from("image1.png")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("image2.png")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("image3.png")` → `b"<content>".to_vec()`
+///
+/// [spec_data]: https://unifiedfontobject.org/versions/ufo3/data/
+/// [spec_images]: https://unifiedfontobject.org/versions/ufo3/images/
+#[derive(Debug, Clone)]
+pub struct Store<T> {
+    items: HashMap<PathBuf, RefCell<Item>>,
+    ufo_root: PathBuf,
+    impl_type: T,
+}
+
+/// Implements custom behavior for the data store.
+#[derive(Debug, Default, Clone)]
+#[doc(hidden)]
+pub struct Data;
+
+pub type DataStore = Store<Data>;
+
+/// Implements custom behavior for the images store.
+#[derive(Debug, Default, Clone)]
+#[doc(hidden)]
+pub struct Image;
+
+pub type ImageStore = Store<Image>;
+
+/// Defines custom behavior for data and images stores.
+#[doc(hidden)]
+pub trait DataType: Default {
+    fn try_list_contents(&self, ufo_root: &Path) -> Result<Vec<PathBuf>, Error>;
+    fn try_load_item(&self, ufo_root: &Path, path: &Path) -> Result<Vec<u8>, StoreError>;
+    fn validate_entry(
+        &self,
+        path: &Path,
+        items: &HashMap<PathBuf, RefCell<Item>>,
+        data: &[u8],
+    ) -> Result<(), StoreError>;
+}
+
+#[derive(Debug, Clone)]
+#[doc(hidden)]
+pub enum Item {
+    NotLoaded,
+    Loaded(Arc<Vec<u8>>),
```

Comment:
minor thing but I would store `Arc<[u8]>` (a reference counted slice) instead of `Arc<Vec<u8>>`. This is minorly more efficient (less pointer indirection, although this is unlikely to matter in practice) but also more clearly expresses that the data cannot be mutated.

Reply:
Ah, ok, so we fully expect people to remove the data they want to mod and then insert it again?